### ISSUE TITLE
Generated swagger creates anonymous structure for post methods with path params

### ISSUE THREAD

**Issue by gbataille** [grpc-ecosystem/grpc-gateway#3706]
## 🐛 Bug Report

For RPC that have a body (typically POST) and one or several path params, the swagger generated uses an anonymous inline schema.
This creates issues with some code generator later on, when generating clients from the swagger definitions

(Arguably the output of protoc-gen-openapiv2 is correct per the spec. What's described here is just very impractical downstream and I think it would be improved by the suggested change)

This is actually visible in this other issue I found https://github.com/grpc-ecosystem/grpc-gateway/issues/2681

## To Reproduce

Take 2 RPCs that use the same input params. But one uses all of them in the body, while the other promotes one to the path

```proto
syntax = "proto3";

import "google/api/annotations.proto";

option go_package = "./test";

service Test{
    rpc AMethod (AMessage) returns (SomeResponse) {
        option (google.api.http) = {
            post: "/v1/projects/someotherpath"
            body: "*"
        };
    }
    rpc BMethod (AMessage) returns (SomeResponse) {
        option (google.api.http) = {
            post: "/v1/projects/someotherpath/{project_id}"
            body: "*"
        };
    }
}
message AMessage {
    string project_id = 1;
    string other_field = 2;
}
message SomeResponse {
    string msg = 1;
}
```

When generating the corresponding swagger
```bash
protoc --plugin=$GOPATH/bin/protoc-gen-openapi --openapiv2_out . --proto_path=$PWD test.proto
```

you get

```json
{
  "swagger": "2.0",
  "info": {
    "title": "test.proto",
    "version": "version not set"
  },
  "tags": [
    {
      "name": "Test"
    }
  ],
  "consumes": [
    "application/json"
  ],
  "produces": [
    "application/json"
  ],
  "paths": {
    "/v1/projects/someotherpath": {
      "post": {
        "operationId": "Test_AMethod",
        "responses": {
          "200": {
            "description": "A successful response.",
            "schema": {
              "$ref": "#/definitions/SomeResponse"
            }
          },
          "default": {
            "description": "An unexpected error response.",
            "schema": {
              "$ref": "#/definitions/rpcStatus"
            }
          }
        },
        "parameters": [
          {
            "name": "body",
            "in": "body",
            "required": true,
            "schema": {
              "$ref": "#/definitions/AMessage"
            }
          }
        ],
        "tags": [
          "Test"
        ]
      }
    },
    "/v1/projects/someotherpath/{projectId}": {
      "post": {
        "operationId": "Test_BMethod",
        "responses": {
          "200": {
            "description": "A successful response.",
            "schema": {
              "$ref": "#/definitions/SomeResponse"
            }
          },
          "default": {
            "description": "An unexpected error response.",
            "schema": {
              "$ref": "#/definitions/rpcStatus"
            }
          }
        },
        "parameters": [
          {
            "name": "projectId",
            "in": "path",
            "required": true,
            "type": "string"
          },
          {
            "name": "body",
            "in": "body",
            "required": true,
            "schema": {
              "type": "object",
              "properties": {
                "otherField": {
                  "type": "string"
                }
              }
            }
          }
        ],
        "tags": [
          "Test"
        ]
      }
    }
  },
  "definitions": {
    "AMessage": {
      "type": "object",
      "properties": {
        "projectId": {
          "type": "string"
        },
        "otherField": {
          "type": "string"
        }
      }
    },
    "SomeResponse": {
      "type": "object",
      "properties": {
        "msg": {
          "type": "string"
        }
      }
    },
    "protobufAny": {
      "type": "object",
      "properties": {
        "@type": {
          "type": "string"
        }
      },
      "additionalProperties": {}
    },
    "rpcStatus": {
      "type": "object",
      "properties": {
        "code": {
          "type": "integer",
          "format": "int32"
        },
        "message": {
          "type": "string"
        },
        "details": {
          "type": "array",
          "items": {
            "type": "object",
            "$ref": "#/definitions/protobufAny"
          }
        }
      }
    }
  }
}

```

What you can see, is that because `AMethod` uses the entire `AMessage` in the body, then the reported schema is
```json
"parameters": [
          {
            "name": "body",
            "in": "body",
            "required": true,
            "schema": {
              "$ref": "#/definitions/AMessage"
            }
          }
        ],
```
which references the proto defined `AMessage` object

However, for `BMethod`, the input proto message is split between the path and the body. This ends up with the swagger using only inline definitions of params (I guess because there is no matching proto-defined object in the definitions)

```json
"parameters": [
          {
            "name": "projectId",
            "in": "path",
            "required": true,
            "type": "string"
          },
          {
            "name": "body",
            "in": "body",
            "required": true,
            "schema": {
              "type": "object",
              "properties": {
                "otherField": {
                  "type": "string"
                }
              }
            }
          }
        ],
```

This, in turn, with some code generator like `./node_modules/.bin/openapi-generator-cli` generates code that looks like this
```ts
public bMethod(projectId: string, body: InlineObject3, observe?: 'body', reportProgress?: boolean, options?: {httpHeaderAccept?: 'application/json',}): Observable<object>;
```

you'll note the use of `InlineObject3` which is not really something we can build upon as this name is based on the order of generation, which might change as the proto evolve

## Expected behavior

The generated swagger should not define body schema inline, but should always used reference to named objects in the `definitions` section

We could use some naming convention like `BMethodBody`. This is what the `go-swagger` code generator actually does it seems.

## Actual Behavior

(Write what happened. Add screenshots, if applicable.)

## Your Environment

```
libprotoc 24.3
PROTOC_GEN_GRPC_GATEWAY_VERSION=v2.18.0
```

**Comment by johanbrandhorst** [comment]
Thanks for the detailed issue! To be clear, is the suggestion that the definitions are good (not populating the path parameters in the body) but that the inlining of the struct is problematic, and you'd prefer a separate definition for it? If so, that makes sense, and I'd be happy to review such a contribution.

**Comment by gbataille** [comment]
Hi @johanbrandhorst 
yes exactly. for me it's great that the proto message gets split in the swagger because it's how it's actually mapped between GRPC and the REST grpc-gateway.

However, indeed, code generator would benefit from having the body part of the argument constituted in a **named** object found in `definitions`

If anybody wants to take a look at it, feel free. I'll try to see what I can do but unfortunately I can't promise anything right now.

**Comment by gbataille** [comment]
Hey @johanbrandhorst 
I wanted to try a bit but
- I don't seem to know how to call the protoc-gen-openapiv2 main. I tried a few things with the `--file` but whatever I'm able to feed in fails.
- I don't seem to be able to build the product with the docker command mentioned in the CONTRIBUTING file 

I can't run it in debug step by step or build it with traces and run it normally, so it's pretty hard for me to figure out how it works and what to change (and test it)

Any pointers?

```
[...]
go: downloading google.golang.org/genproto v0.0.0-20231016165738-49dd2c1f3d0b
find . -type f -name '*.pb.go' -delete
find . -type f -name '*.swagger.json' -delete
find . -type f -name '*.pb.gw.go' -delete
rm -f examples/internal/clients/echo/client.go examples/internal/clients/echo/response.go examples/internal/clients/echo/configuration.go examples/internal/clients/echo/api_echo_service.go examples/internal/clients/echo/model_examplepb_simple_message.go examples/internal/clients/echo/model_examplepb_embedded.go examples/internal/clients/abe/model_a_bit_of_everything_nested.go examples/internal/clients/abe/api_a_bit_of_everything_service.go examples/internal/clients/abe/client.go examples/internal/clients/abe/api_camel_case_service_name.go examples/internal/clients/abe/configuration.go examples/internal/clients/abe/api_echo_rpc.go examples/internal/clients/abe/model_examplepb_a_bit_of_everything.go examples/internal/clients/abe/model_examplepb_a_bit_of_everything_repeated.go examples/internal/clients/abe/model_examplepb_body.go examples/internal/clients/abe/model_examplepb_numeric_enum.go examples/internal/clients/abe/model_examplepb_update_v2_request.go examples/internal/clients/abe/model_message_path_enum_nested_path_enum.go examples/internal/clients/abe/model_nested_deep_enum.go examples/internal/clients/abe/model_pathenum_path_enum.go examples/internal/clients/abe/model_protobuf_field_mask.go examples/internal/clients/abe/response.go examples/internal/clients/unannotatedecho/client.go examples/internal/clients/unannotatedecho/response.go examples/internal/clients/unannotatedecho/configuration.go examples/internal/clients/unannotatedecho/model_examplepb_unannotated_simple_message.go examples/internal/clients/unannotatedecho/api_unannotated_echo_service.go examples/internal/clients/responsebody/client.go examples/internal/clients/responsebody/response.go examples/internal/clients/responsebody/configuration.go examples/internal/clients/responsebody/model_examplepb_repeated_response_body_out.go examples/internal/clients/responsebody/model_examplepb_repeated_response_body_out_response.go examples/internal/clients/responsebody/model_examplepb_repeated_response_strings.go examples/internal/clients/responsebody/model_examplepb_response_body_out.go examples/internal/clients/responsebody/model_examplepb_response_body_out_response.go examples/internal/clients/responsebody/model_response_response_type.go examples/internal/clients/responsebody/api_response_body_service.go examples/internal/clients/generateunboundmethods/client.go examples/internal/clients/generateunboundmethods/response.go examples/internal/clients/generateunboundmethods/configuration.go examples/internal/clients/generateunboundmethods/model_examplepb_generate_unbound_methods_simple_message.go examples/internal/clients/generateunboundmethods/api_generate_unbound_methods.go
# These generation steps are run in order so that later steps can
# overwrite files produced by previous steps, if necessary.
buf generate
google/api/http.proto:29:9:symbol "google.api.Http" already defined at 3706_test/google/api/http.proto:29:9
google/api/http.proto:33:21:symbol "google.api.Http.rules" already defined at 3706_test/google/api/http.proto:33:21
[...]
google/protobuf/descriptor.proto:914:20:symbol "google.protobuf.GeneratedCodeInfo.Annotation.begin" already defined at 3706_test/google/protobuf/descriptor.proto:914:20
google/protobuf/descriptor.proto:919:20:symbol "google.protobuf.GeneratedCodeInfo.Annotation.end" already defined at 3706_test/google/protobuf/descriptor.proto:919:20
make: *** [Makefile:93: proto] Error 100
```

**Comment by johanbrandhorst** [comment]
The easiest way to "integration test" the generator locally is using `protoc` or `buf`. For example, you can run `make install` and `make proto` to regenerate all the files locally. If you've made changes to `protoc-gen-openapiv2`, they will be reflected in the generated swagger files.

**Comment by johanbrandhorst** [comment]
What command did you run to get that error? I don't know where `3706_test` comes from.

**Comment by gbataille** [comment]
> The easiest way to "integration test" the generator locally is using `protoc` or `buf`. For example, you can run `make install` and `make proto` to regenerate all the files locally. If you've made changes to `protoc-gen-openapiv2`, they will be reflected in the generated swagger files.

yes that's one of the way I wanted to do that but using the docker command found in the `CONTRIBUTING` file to build the product (supposedly with all the dependencies in the build image I guess), I got the build error.

so sorry let me rephrase. I tried 2 things
- first I wanted to try and debug the protoc-gen-openapiv2 step by step, by using the main.go that is there. That I was not able to do. I can't seem to use the main entrypoint. I don't know if/how it's supposed to work
- for the error message I pasted, I used the docker command that is in the CONTRIBUTING file. but it's funny you mention this 3706_test. it is indeed a folder/proto file that I created in the project folder, to play with. I did not spot it in the long stack. I edited it and copied only a part here and was not careful what was there. This feels like a warning more than an error, but possibly this is the reason why the build command fail, so I guess I'll retry without this new folder

Thanks for spotting that
